File: backend/services/analyzer.py
"""Celebrity name extraction and trending analysis."""
from __future__ import annotations
from typing import Callable, List, Optional
from openai import OpenAI
from models.schemas import PostItem


def get_trending_celebs(
    posts: List[PostItem],
    client: OpenAI,
    top_n: int = 5,
    on_progress: Optional[Callable[[int, int], None]] = None,
) -> List[str]:
    if not posts:
        return []

    titles_block = "\n".join(f"{i+1}. {p.title}" for i, p in enumerate(posts))

    if on_progress:
        on_progress(30, 100)

    resp = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[
            {
                "role": "system",
                "content": (
                    "당신은 연예 트렌드 분석 전문가입니다. 제공된 블로그 게시글 제목 목록을 분석하여 "
                    "현재 가장 많이 언급되거나 화제가 되고 있는 연예인(셀럽)들을 추출하세요.\n"
                    "규칙:\n"
                    "1. 제목에서 옷, 가방, 액세서리 정보가 포함된 연예인을 우선순위로 두세요.\n"
                    "2. 상위 연예인들의 이름을 중요도 순으로 나열하세요.\n"
                    "3. 이름만 쉼표(,)로 구분하여 정확히 출력하세요. 부가 설명은 생략하세요."
                ),
            },
            {"role": "user", "content": f"분석할 제목 목록:\n{titles_block}"},
        ],
        temperature=0.3,
    )

    if on_progress:
        on_progress(90, 100)

    content = (resp.choices[0].message.content or "").strip()
    celebs = [
        c.strip().lstrip("0123456789. ")
        for c in content.replace("\n", ",").split(",")
    ]
    celebs = [c for c in celebs if 1 < len(c) < 12]
    # Deduplicate preserving order
    seen: set[str] = set()
    unique = [c for c in celebs if not (c in seen or seen.add(c))]  # type: ignore

    if on_progress:
        on_progress(100, 100)

    return unique[:top_n]

File: standalone/src/lib/analyzer.ts
import OpenAI from "openai";
import type { PostItem } from "./collector";

export function createClient(apiKey: string): OpenAI {
  return new OpenAI({ apiKey, dangerouslyAllowBrowser: true });
}

export async function extractCeleb(title: string, client: OpenAI): Promise<string> {
  try {
    const res = await client.chat.completions.create({
      model: "gpt-4o-mini",
      messages: [
        {
          role: "system",
          content:
            "문구를 입력하면 그 문구로부터 연예인의 이름을 찾아서 출력해줘. " +
            "문장이 나와도 문장에 대답하지 말고, 해당 문장에 등장하는 연예인의 이름만 찾아서 출력해. " +
            "여러 명이면 쉼표로 구분해. 연예인이 없으면 빈 문자열만 출력해.",
        },
        { role: "user", content: "이달의소녀덤덤이달의소녀 롤링다이스로 롤아웃 풀세트 연봄빛 코드" },
        { role: "assistant", content: "이달의소녀" },
        { role: "user", content: "이도경5이도경이봉지아 1회 이도경의 봄 봄트 착용 의상" },
        { role: "assistant", content: "이도경, 봉지아" },
        { role: "user", content: "이재호 혼자사는 이민호 집 거실 뭐가있나요?" },
        { role: "assistant", content: "이재호" },
        { role: "user", content: title },
      ],
      temperature: 1,
      max_tokens: 100,
    });
    return res.choices[0].message.content?.trim() ?? "";
  } catch {
    return "";
  }
}

export async function getTrendingCelebs(
  posts: PostItem[],
  client: OpenAI,
  topN = 3,
  onProgress?: (done: number, total: number) => void
): Promise<string[]> {
  const counts = new Map<string, number>();

  for (let i = 0; i < posts.length; i++) {
    const result = await extractCeleb(posts[i].title, client);
    result
      .split(",")
      .map((c) => c.trim())
      .filter(Boolean)
      .forEach((c) => counts.set(c, (counts.get(c) ?? 0) + 1));
    onProgress?.(i + 1, posts.length);
  }

  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, topN)
    .map(([name]) => name);
}
